crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -269,48 +271,244 @@ impl ClassInfoConstraintFunction {
     db: &'db dyn Db,
 ) {
-    for (key, value) in from {
-        match into.entry(*key) {
+    for (key, from_constraint) in &from.constraints {
+        match into.constraints.entry(*key) {
             Entry::Occupied(mut entry) => {
-                *entry.get_mut() = IntersectionBuilder::new(db)
-                    .add_positive(*entry.get())
-                    .add_positive(*value)
-                    .build();
+                let into_constraint = entry.get();
+
+                // Distribute AND over OR: (A1 | A2 | ...) AND (B1 | B2 | ...)
+                // becomes (A1 & B1) | (A1 & B2) | ... | (A2 & B1) | ...
+                let mut new_disjuncts = SmallVec::new();
+
+                for left_conj in &into_constraint.disjuncts {
+                    for right_conj in &from_constraint.disjuncts {
+                        if right_conj.typeguard.is_some() {
+                            // If the right conjunct has a TypeGuard, it "wins" the conjunction
+                            new_disjuncts.push(*right_conj);
+                        } else {
+                            // Intersect the regular constraints
+                            let new_regular = IntersectionBuilder::new(db)
+                                .add_positive(left_conj.constraint)
+                                .add_positive(right_conj.constraint)
+                                .build();
+
+                            new_disjuncts.push(Conjunction {
+                                constraint: new_regular,
+                                typeguard: left_conj.typeguard,
+                            });
+                        }
+                    }
+                }
+
+                *entry.get_mut() = NarrowingConstraint {
+                    disjuncts: new_disjuncts,
+                };
             }
             Entry::Vacant(entry) => {
-                entry.insert(*value);
+                entry.insert(from_constraint.clone());
             }
         }
     }
 }
 
+/// Merge constraints with OR semantics (union/disjunction).
+///
+/// When we have `constraint1 OR constraint2`, we simply concatenate the disjuncts
+/// from both constraints: `(A | B) OR (C | D)` becomes `A | B | C | D`
+///
+/// However, if a place appears in only one branch of the OR, we need to widen it
+/// to `object` in the overall result (because the other branch doesn't constrain it).
 fn merge_constraints_or<'db>(
-    into: &mut NarrowingConstraints<'db>,
-    from: &NarrowingConstraints<'db>,
-    db: &'db dyn Db,
+    into: &mut InternalConstraints<'db>,
+    from: &InternalConstraints<'db>,
+    _db: &'db dyn Db,
 ) {
-    for (key, value) in from {
-        match into.entry(*key) {
+    // For places that appear in `into` but not in `from`, widen to object
+    for (key, value) in &mut into.constraints {
+        if !from.constraints.contains_key(key) {
+            *value = NarrowingConstraint::regular(Type::object());
+        }
+    }
```

Comment:
This seems better, and passes tests, not sure why we didn't do it previously:

```suggestion
    // For places that appear in `into` but not in `from`, do not constrain
    into.constraints
        .retain(|key, _| from.constraints.contains_key(key));

```